### PR TITLE
feat(perpetual-liquidatable) Implement Liquidatable and replace expiry logic with emergency-shutdown logic

### LINKED ISSUE #2048: Liquidatable: remove modifiers that reference expiration (but still handle emergency shutdown).
**Why**

Project Evergreen.

**Difficulty Score [1-10]**

- 5: ~4 hours

### PR DESCRIPTION
**Pull request by nicholaspai**
**Motivation**

#2024, #2048

**Summary**

This PR copies the ExpiringMultiparty's Liquidatable contract and strips out expiry logic. This contract imports all of the logic from the PerpetualPositionManager introduced in #2072. Unit tests were also copied from ExpiringMultiparty and minimal changes were made to replace expiry logic with emergency shutdown logic.

**Details**

Major changes:
- Add `_getOraclePrice(timestamp)` back to `PerpetualPositionManager` because the Liquidatable contract needs to call it on the liquidation-timestamp.
- `createLiquidation` should be blocked during an emergency shutdown, for the EMP it is blocked post-expiry.

**Issue(s)**

Initial implementation of #2024 
close #2048 

A Follow up PR will change the `_settle` logic to incorporate a cumulativeFundingRateMultiplier, but this is likely blocked until an initial implementation for #2030 is built.

**Open Questions**
Is there any reason to disable Disputes and withdrawals of liquidation/dispute rewards following an emergency shutdown (for the perpetual contract)?

- Cons to Disabling: If someone gets liquidated incorrectly right before shutdown, then they cannot recover their funds
- Pros: Guarantees that no collateral leaves the contract in a shutdown

I’m inclined to follow the EMP pattern and allow disputes/withdrawal of liquidation/dispute rewards. This implies that once a sponsor is liquidated, their collateral has already “left the contract” so we should not freeze these funds

### CODE REVIEW

File: packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol [hunk 1]
```diff
@@ -719,12 +719,11 @@ contract PerpetualPositionManager is FeePayer, AdministrateeInterface {
     }
 
     // Fetches a resolved Oracle price from the Oracle. Reverts if the Oracle hasn't resolved for this request.
-    function _getOracleEmergencyShutdownPrice() internal view returns (FixedPoint.Unsigned memory) {
+    function _getOraclePrice(uint256 requestedTime) internal view returns (FixedPoint.Unsigned memory) {
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol:722]: Liquidatable needs access to this internal method to get the price for a specific liquidation time

File: packages/core/test/financial-templates/perpetual-multiparty/Liquidatable.js [hunk 2]
```diff
@@ -0,0 +1,2050 @@
...
+          // State should be uninitialized as the struct has been deleted as a result of the withdrawal.
+          // Once a dispute fails and the liquidator withdraws the struct is removed from state.
+          ev.liquidationStatus.toString() == LiquidationStatesEnum.UNINITIALIZED &&
+          ev.settlementPrice.toString() == disputePrice.toString()
+        );
+      });
+    });
+  });
+
+  describe("Emergency shutdown", () => {
```
Comment (nicholaspai) [inline packages/core/test/financial-templates/perpetual-multiparty/Liquidatable.js:1600]: This test and the next one replace expiry logic with emergency-shutdown logic

File: packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol (hunk 1 above)
Comment (chrismaree) [reply packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol:722]: that's a good point. needs to get it at dispute time.

**Review by chrismaree — APPROVED** [review]
Great PR @nicholaspai! very simple in that we basically had to change only one modifier to get this to work with the perpetual contract. Tests are also 99.9% the same as before except for one small refactor. I'm happy with it :)

**Comment by chrismaree** [comment]
Regarding your open question, as discussed offline, I think it best to keep the same EMP logic as a sponsor could be incorrectly liquidated right before the contract is emergency shutdown.

**Review by mrice32 — COMMENTED** [review]
@nicholaspai thoughts on copying the contract over in a separate PR? That way, we can tell what parts changed and need to be reviewed? Otherwise, we'll probably miss the parts you changed.

**Comment by nicholaspai** [comment]
> @nicholaspai thoughts on copying the contract over in a separate PR? That way, we can tell what parts changed and need to be reviewed? Otherwise, we'll probably miss the parts you changed.

Normally i'd say yes, but there are very few changes. I'd recommend using a [diff checker tool](https://www.diffchecker.com/) to see that I only removed the `expirationTimestamp` from the constructor and replaced the one `onlyPreExpiry()` modifier in front of `createLiquidation()` with `notEmergencyShutdown`

**Review by daywiss — APPROVED** [review]

**Review by mrice32 — APPROVED** [review]

**Pull request merged**